zone/spells.cpp

```diff
@@ -3619,45 +3576,30 @@ int Mob::AddBuff(Mob *caster, uint16 spell_id, int duration, int32 level_overrid
 						);
 					}
 
-					const bool caster_has_block_event = (
-						(caster->IsBot() && parse->BotHasQuestSub(EVENT_SPELL_BLOCKED)) ||
-						(caster->IsClient() && parse->PlayerHasQuestSub(EVENT_SPELL_BLOCKED)) ||
-						(caster->IsNPC() && parse->HasQuestSub(caster->GetNPCTypeID(), EVENT_SPELL_BLOCKED))
-					);
-
-					const bool cast_on_has_block_event = (
-						(IsBot() && parse->BotHasQuestSub(EVENT_SPELL_BLOCKED)) ||
-						(IsClient() && parse->PlayerHasQuestSub(EVENT_SPELL_BLOCKED)) ||
-						(IsNPC() && parse->HasQuestSub(GetNPCTypeID(), EVENT_SPELL_BLOCKED))
-					);
-
-					if (caster_has_block_event || cast_on_has_block_event) {
-						const std::string& export_string = fmt::format(
-							"{} {}",
-							curbuf.spellid,
-							spell_id
-						);
-
-						if (caster_has_block_event) {
-							if (caster->IsBot()) {
-								parse->EventBot(EVENT_SPELL_BLOCKED, caster->CastToBot(), this, export_string, 0);
-							} else if (caster->IsClient()) {
-								parse->EventPlayer(EVENT_SPELL_BLOCKED, caster->CastToClient(), export_string, 0);
-							} else if (caster->IsNPC()) {
-								parse->EventNPC(EVENT_SPELL_BLOCKED, caster->CastToNPC(), this, export_string, 0);
-							}
+					parse->EventPlayerNPCBotMerc(
+						EVENT_SPELL_BLOCKED,
+						caster,
+						this,
+						[&]() {
+							return fmt::format(
+								"{} {}",
+								curbuf.spellid,
+								spell_id
+							);
 						}
+					);
 
-						if (cast_on_has_block_event && caster != this) {
-							if (IsBot()) {
-								parse->EventBot(EVENT_SPELL_BLOCKED, CastToBot(), caster, export_string, 0);
-							} else if (IsClient()) {
-								parse->EventPlayer(EVENT_SPELL_BLOCKED, CastToClient(), export_string, 0);
-							} else if (IsNPC()) {
-								parse->EventNPC(EVENT_SPELL_BLOCKED, CastToNPC(), caster, export_string, 0);
-							}
+					parse->EventPlayerNPCBotMerc(
+						EVENT_SPELL_BLOCKED,
+						this,
+						[&]() {
+							return fmt::format(
+								"{} {}",
+								curbuf.spellid,
+								spell_id
+							);
 						}
-					}
+					);
```

Comment:
This call is missing the `caster` argument for initiator that the old code had. It's also missing the  `if (caster != this)` check that prevented the caster from getting double events if casting on self

The duplicate lambda could also be re-used to turn both EVENT_SPELL_BLOCKED calls into one liners.